# Bank Customer Segmentation

# 01 Importing Libraries

In [2]:
import pandas as pd
import numpy as np
import os
from datetime import datetime


# 02 Importing Dataset

In [2]:
path=r'C:\Users\sruth\OneDrive\Desktop\Career Foundry\Data Immersion\A6-Advanced Analytics & Dashboard design\6.1 Sourcing open data'

In [3]:
df_bank=pd.read_csv(os.path.join(path,'Data','Original Data','bank_transactions.csv'))

# 03 Exploratory Data Analysis

In [4]:
df_bank.shape

(1048567, 9)

In [5]:
df_bank.head(10)

TransactionID CustomerID CustomerDOB CustGender CustLocation  \
0            T1   C5841053     10/1/94          F   JAMSHEDPUR   
1            T2   C2142763      4/4/57          M      JHAJJAR   
2            T3   C4417068    26/11/96          F       MUMBAI   
3            T4   C5342380     14/9/73          F       MUMBAI   
4            T5   C9031234     24/3/88          F  NAVI MUMBAI   
5            T6   C1536588     8/10/72          F     ITANAGAR   
6            T7   C7126560     26/1/92          F       MUMBAI   
7            T8   C1220223     27/1/82          M       MUMBAI   
8            T9   C8536061     19/4/88          F      GURGAON   
9           T10   C6638934     22/6/84          M       MUMBAI   

   CustAccountBalance TransactionDate  TransactionTime  \
0            17819.05          2/8/16           143207   
1             2270.69          2/8/16           141858   
2            17874.44          2/8/16           142712   
3           866503.21          2/8/16           142714   
4             6714.43          2/8/16           181156   
5            53609.20          2/8/16           173940   
6              973.46          2/8/16           173806   
7            95075.54          2/8/16           170537   
8            14906.96          2/8/16           192825   
9             4279.22          2/8/16           192446   

   TransactionAmount (INR)  
0                    25.00  
1                 27999.00  
2                   459.00  
3                  2060.00  
4                  1762.50  
5                   676.00  
6                   566.00  
7                   148.00  
8                   833.00  
9                   289.11

In [6]:
df_bank.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048567 entries, 0 to 1048566
Data columns (total 9 columns):
 #   Column                   Non-Null Count    Dtype  
---  ------                   --------------    -----  
 0   TransactionID            1048567 non-null  object 
 1   CustomerID               1048567 non-null  object 
 2   CustomerDOB              1045170 non-null  object 
 3   CustGender               1047467 non-null  object 
 4   CustLocation             1048416 non-null  object 
 5   CustAccountBalance       1046198 non-null  float64
 6   TransactionDate          1048567 non-null  object 
 7   TransactionTime          1048567 non-null  int64  
 8   TransactionAmount (INR)  1048567 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 72.0+ MB


In [7]:
# As mentioned by author,TransactionTime is in Unix timestamp.So changing to readable time format
# df_bank['TransactionTime'] =pd.to_datetime(df_bank[‘TransactionTime’], unit='s')
# not in Unix timestamp

In [8]:
# TransactionTime is in HHMMSS format.So changing to readable time format
df_bank['TransactionTime'] = df_bank['TransactionTime'].apply(lambda x : datetime.utcfromtimestamp(int(x)).strftime('%H:%M:%S'))

In [9]:
df_bank.head(10)

TransactionID CustomerID CustomerDOB CustGender CustLocation  \
0            T1   C5841053     10/1/94          F   JAMSHEDPUR   
1            T2   C2142763      4/4/57          M      JHAJJAR   
2            T3   C4417068    26/11/96          F       MUMBAI   
3            T4   C5342380     14/9/73          F       MUMBAI   
4            T5   C9031234     24/3/88          F  NAVI MUMBAI   
5            T6   C1536588     8/10/72          F     ITANAGAR   
6            T7   C7126560     26/1/92          F       MUMBAI   
7            T8   C1220223     27/1/82          M       MUMBAI   
8            T9   C8536061     19/4/88          F      GURGAON   
9           T10   C6638934     22/6/84          M       MUMBAI   

   CustAccountBalance TransactionDate TransactionTime  TransactionAmount (INR)  
0            17819.05          2/8/16        15:46:47                    25.00  
1             2270.69          2/8/16        15:24:18                 27999.00  
2            17874.44          2/8/16        15:38:32                   459.00  
3           866503.21          2/8/16        15:38:34                  2060.00  
4             6714.43          2/8/16        02:19:16                  1762.50  
5            53609.20          2/8/16        00:19:00                   676.00  
6              973.46          2/8/16        00:16:46                   566.00  
7            95075.54          2/8/16        23:22:17                   148.00  
8            14906.96          2/8/16        05:33:45                   833.00  
9             4279.22          2/8/16        05:27:26                   289.11

In [10]:
# to find unique custlocation
locations= df_bank['CustLocation'].unique()

In [11]:
len(locations)

9356

In [12]:
#To find count(custlocation) groupby custlocation
location_count= df_bank.groupby('CustLocation')['CustLocation'].count()


In [13]:
location_count

CustLocation
(154) BHASKOLA FARIDABAD          11
(BEFORE YMCA BLDG) CHENNAI         1
(BENAKA MDTS) BANGALORE           21
(BRINDA BAN ) KOLKATA              5
(DT) HOSUR                        11
                                  ..
ZOPADPATTI TEHSIL KURLA MUMBAI     1
ZUARINAGAR                         6
ZULFA DAIRY MUMBAI                 4
ZUMARI TILAIYA                     9
ZUNHEBOTO                          1
Name: CustLocation, Length: 9355, dtype: int64

In [14]:
#To find count(custlocation) groupby customerlocation having locations with no of transactions >5000
loc_count_5000=location_count[location_count>5000]

In [15]:
loc_count_5000

CustLocation
AHMEDABAD       12264
BANGALORE       81555
CHANDIGARH       9526
CHENNAI         30009
DELHI           71019
FARIDABAD       11318
GHAZIABAD       15834
GURGAON         73818
HYDERABAD       23049
JAIPUR           9921
KOLKATA         19974
LUCKNOW          7763
MOHALI           6741
MUMBAI         103595
NAVI MUMBAI     13080
NEW DELHI       84928
NOIDA           32784
PUNE            25851
SURAT            5770
THANE           21505
Name: CustLocation, dtype: int64

# Create a subset of dataset with Custlocation having no of transaction >5000

In [16]:
# Create a subset of dataset with Custlocation having no of transaction >5000
df_bank_5000= df_bank.groupby("CustLocation").filter(lambda x: len(x) > 5000)

In [17]:
df_bank_5000

TransactionID CustomerID CustomerDOB CustGender CustLocation  \
2                  T3   C4417068    26/11/96          F       MUMBAI   
3                  T4   C5342380     14/9/73          F       MUMBAI   
4                  T5   C9031234     24/3/88          F  NAVI MUMBAI   
6                  T7   C7126560     26/1/92          F       MUMBAI   
7                  T8   C1220223     27/1/82          M       MUMBAI   
...               ...        ...         ...        ...          ...   
1048560      T1048561   C5028150     24/5/80          M         PUNE   
1048561      T1048562   C1034220    19/11/92          M    BANGALORE   
1048562      T1048563   C8020229      8/4/90          M    NEW DELHI   
1048564      T1048565   C6412354     18/5/89          M    HYDERABAD   
1048566      T1048567   C8337524      5/3/84          M         PUNE   

         CustAccountBalance TransactionDate TransactionTime  \
2                  17874.44          2/8/16        15:38:32   
3                 866503.21          2/8/16        15:38:34   
4                   6714.43          2/8/16        02:19:16   
6                    973.46          2/8/16        00:16:46   
7                  95075.54          2/8/16        23:22:17   
...                     ...             ...             ...   
1048560              464.87         18/9/16        03:17:12   
1048561             1212.70         18/9/16        03:18:54   
1048562             7635.19         18/9/16        03:20:24   
1048564           221757.06         18/9/16        02:55:13   
1048566            75734.42         18/9/16        02:20:22   

         TransactionAmount (INR)  
2                          459.0  
3                         2060.0  
4                         1762.5  
6                          566.0  
7                          148.0  
...                          ...  
1048560                   3000.0  
1048561                   1500.0  
1048562                    799.0  
1048564                    770.0  
1048566                   1166.0  

[660304 rows x 9 columns]

In [18]:
# To check the accuracy of subset
loc_count_5000.sum()

660304

In [19]:
# Locations in column ‘CustLocation’ were modified to follow uniform naming practices
df_bank_5000['CustLocation'] = df_bank_5000['CustLocation'].replace(['DELHI','NAVI MUMBAI'], ['NEW DELHI','MUMBAI'])

In [20]:
df_bank_5000

TransactionID CustomerID CustomerDOB CustGender CustLocation  \
2                  T3   C4417068    26/11/96          F       MUMBAI   
3                  T4   C5342380     14/9/73          F       MUMBAI   
4                  T5   C9031234     24/3/88          F       MUMBAI   
6                  T7   C7126560     26/1/92          F       MUMBAI   
7                  T8   C1220223     27/1/82          M       MUMBAI   
...               ...        ...         ...        ...          ...   
1048560      T1048561   C5028150     24/5/80          M         PUNE   
1048561      T1048562   C1034220    19/11/92          M    BANGALORE   
1048562      T1048563   C8020229      8/4/90          M    NEW DELHI   
1048564      T1048565   C6412354     18/5/89          M    HYDERABAD   
1048566      T1048567   C8337524      5/3/84          M         PUNE   

         CustAccountBalance TransactionDate TransactionTime  \
2                  17874.44          2/8/16        15:38:32   
3                 866503.21          2/8/16        15:38:34   
4                   6714.43          2/8/16        02:19:16   
6                    973.46          2/8/16        00:16:46   
7                  95075.54          2/8/16        23:22:17   
...                     ...             ...             ...   
1048560              464.87         18/9/16        03:17:12   
1048561             1212.70         18/9/16        03:18:54   
1048562             7635.19         18/9/16        03:20:24   
1048564           221757.06         18/9/16        02:55:13   
1048566            75734.42         18/9/16        02:20:22   

         TransactionAmount (INR)  
2                          459.0  
3                         2060.0  
4                         1762.5  
6                          566.0  
7                          148.0  
...                          ...  
1048560                   3000.0  
1048561                   1500.0  
1048562                    799.0  
1048564                    770.0  
1048566                   1166.0  

[660304 rows x 9 columns]

In [21]:
df_bank_5000['CustLocation'].value_counts(dropna=False)

NEW DELHI     155947
MUMBAI        116675
BANGALORE      81555
GURGAON        73818
NOIDA          32784
CHENNAI        30009
PUNE           25851
HYDERABAD      23049
THANE          21505
KOLKATA        19974
GHAZIABAD      15834
AHMEDABAD      12264
FARIDABAD      11318
JAIPUR          9921
CHANDIGARH      9526
LUCKNOW         7763
MOHALI          6741
SURAT           5770
Name: CustLocation, dtype: int64

In [22]:
df_bank_5000.describe()

CustAccountBalance  TransactionAmount (INR)
count        6.588410e+05             6.603040e+05
mean         1.327436e+05             1.623003e+03
std          8.898709e+05             6.966071e+03
min          0.000000e+00             0.000000e+00
25%          5.703030e+03             2.000000e+02
50%          2.000938e+04             5.000000e+02
75%          6.847865e+04             1.280000e+03
max          8.224463e+07             1.560035e+06

In [23]:
# To check any missing values
df_bank_5000.isnull().sum()

TransactionID                 0
CustomerID                    0
CustomerDOB                2405
CustGender                 1007
CustLocation                  0
CustAccountBalance         1463
TransactionDate               0
TransactionTime               0
TransactionAmount (INR)       0
dtype: int64

In [24]:
df_bank_5000['CustomerDOB'].value_counts(dropna=False)

1/1/1800    36367
NaN          2405
1/1/89        501
1/7/90        458
1/1/90        390
            ...  
30/5/63         1
20/3/52         1
8/2/51          1
4/8/44          1
18/7/65         1
Name: CustomerDOB, Length: 15137, dtype: int64

# There are 2405 missing values  and 36367 wrong customer DOB(1/1/1800) in column 'CustomerDOB'

In [25]:
df_bank_5000['CustGender'].value_counts(dropna=False)

M      466663
F      192633
NaN      1007
T           1
Name: CustGender, dtype: int64

# There are 1007 missing values in column 'CustGender'

In [26]:
df_bank_5000['CustAccountBalance'].value_counts(dropna=False)

0.00        1682
NaN         1463
27021.05     514
25256.28     295
25525.44     212
            ... 
28406.66       1
11422.97       1
8400.38        1
79546.56       1
15037.89       1
Name: CustAccountBalance, Length: 97285, dtype: int64

# There are 1463 Missing values in column 'CustAccountBalance'

In [27]:
# The missing values in the columns (CustomerDOB, CustGender) may be the result of customer not providing the information, 
# while the column (CustAccountBalance) may be due to zero balance.
# So I am imputing missing values in DOB and Gender with ‘NA’ and in CustAccountBalance with ‘0’.

In [28]:
df_bank_5000['CustomerDOB'].fillna('NA',inplace=True)

In [29]:
df_bank_5000['CustomerDOB'] = df_bank_5000['CustomerDOB'].replace(['1/1/1800'], ['NA'])

In [30]:
df_bank_5000['CustomerDOB'].value_counts(dropna=False)

NA         38772
1/1/89       501
1/7/90       458
1/1/90       390
12/7/78      378
           ...  
30/5/63        1
20/3/52        1
8/2/51         1
4/8/44         1
18/7/65        1
Name: CustomerDOB, Length: 15136, dtype: int64

In [31]:
df_bank_5000['CustGender'].fillna('NA',inplace=True)

In [32]:
df_bank_5000['CustGender'].value_counts(dropna=False)

M     466663
F     192633
NA      1007
T          1
Name: CustGender, dtype: int64

In [33]:
df_bank_5000['CustAccountBalance'].fillna(0,inplace=True)

In [34]:
df_bank_5000['CustAccountBalance'].value_counts(dropna=False)

0.00        3145
27021.05     514
25256.28     295
25525.44     212
84098.47     198
            ... 
22576.45       1
24543.35       1
70112.21       1
2582.29        1
15037.89       1
Name: CustAccountBalance, Length: 97284, dtype: int64

In [35]:
# To check is there any duplicate data
df_bank_5000[df_bank_5000.duplicated()]

Empty DataFrame
Columns: [TransactionID, CustomerID, CustomerDOB, CustGender, CustLocation, CustAccountBalance, TransactionDate, TransactionTime, TransactionAmount (INR)]
Index: []

# There are no duplicates record in dataset

In [36]:
# To check is there any mixed datatype
for col in df_bank_5000.columns.tolist():
    weird = (df_bank_5000[[col]].applymap(type) != df_bank_5000[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_bank_5000[weird]) > 0:
        print (col)

# There are no mixed datatype columns

# 04 Exporting Dataset

In [37]:
df_bank_5000.to_pickle(os.path.join(path,'Data','Prepared Data','Bank Transactions.pkl'))

In [42]:
df_bank_5000.to_csv(os.path.join(path,'Data','Prepared Data','Bank Transactions.csv'),index=False)